### DISCUSSION TITLE
How to make SSH work on your IntelliJ after setting it up

### DISCUSSION THREAD

**Discussion by MvuleniNofemele** [community/community#64472] [New to GitHub]
I'm still getting authentication error message when I try to commit or push code to Github, how do I get to push my code to Github withouth using the old method of password request on my Terminal.

Please see screenshot below

![image](https://github.com/community/community/assets/133951877/892ca2b3-b58a-4066-8805-7958abbb1145)

**Comment by w8ste** [comment]
@MvuleniNofemele First of all, you need set up an ssh-connections. [Here](https://docs.github.com/en/authentication/connecting-to-github-with-ssh) is how to do that. The remote repository is still set to the HTTPS-Address. Please try running the following commands (im assuming the remote repo, is called origin):
```
git remote remove origin
git remote add origin "ssh"
```
Replace "ssh" with the ssh-url of your remote Repository, which you can find here:
![image](https://github.com/community/community/assets/115741870/073cfe9d-c3bc-4ffd-800f-793cad8a2162)

Afterwards it should work. Hope this helps :)

**Comment by arafajs** [comment]
To push your code to GitHub without using the old method of password request on your terminal, you can use a personal access token (PAT). A PAT is a unique string of characters that you can use to authenticate yourself to GitHub without having to enter your password every time.

To create a PAT, follow these steps:

1. Go to your GitHub account settings.
2. Click on "Developer settings".
3. Click on "Personal access tokens".
4. Click on "Generate new token".
5. Give your token a name and select the scopes that you need.
6. Click on "Generate token".

Once you have created a PAT, you can use it to push your code to GitHub by following these steps:

- Open a terminal window and navigate to the directory that contains your local Git repository.
- Run the following command to add your PAT to your SSH configuration:

`git config --global credential.helper ssh-agent`

- Run the following command to load your PAT into the SSH agent:

`ssh-agent bash`

- Paste your PAT into the prompt.
- Press Enter.

Run the following command to push your code to GitHub:
`git push origin master`
If you are still getting an authentication error, make sure that you have entered your PAT correctly. You can also try restarting your terminal and your computer.

**Reply by MvuleniNofemele** [reply]
Hi @arafajs  I'm getting this message on Terminal after hitting enter

C:\........>ssh-agent bash
unable to start ssh-agent service, error :1058

**Reply by MvuleniNofemele** [reply]
C:\.............>git push origin master
fatal: 'ssh' does not appear to be a git repository
fatal: Could not read from remote repository.

**Comment by Jaber-Saed** [comment]
If you're encountering authentication errors when trying to commit or push code to GitHub, it's likely due to changes in GitHub's authentication mechanisms. GitHub now strongly recommends using personal access tokens (PATs) or SSH keys for authentication instead of using passwords for security reasons. Here's how you can set up either of these methods to push your code without using the old password-based authentication:

### Option 1: Personal Access Token (PAT)

1. **Generate a Personal Access Token:**
   - Go to your GitHub account settings.
   - In the left sidebar, click on "Developer settings" > "Personal access tokens."
   - Click on "Generate new token."
   - Give your token a name, select the desired scopes (permissions), and click "Generate token."
   - Copy the generated token. **Note:** You won't be able to see this token again, so make sure to store it securely.

2. **Use the Personal Access Token:**
   - In your terminal, when prompted for credentials during a push or clone operation, use your username and the generated personal access token as your password.

### Option 2: SSH Key

1. **Check for Existing SSH Key:**
   - Check if you already have an SSH key on your local machine. By default, the key pair is usually located in `~/.ssh/id_rsa` (private key) and `~/.ssh/id_rsa.pub` (public key).

2. **Generate an SSH Key (If Needed):**
   - If you don't have an SSH key, you can generate one using the following command:
     ```sh
     ssh-keygen -t rsa -b 4096 -C "your_email@example.com"
     ```
   - Follow the prompts to save the key pair in the default location or specify a custom location.

3. **Add SSH Public Key to GitHub:**
   - Go to your GitHub account settings.
   - In the left sidebar, click on "SSH and GPG keys."
   - Click on "New SSH key."
   - Copy the contents of your public key file (`id_rsa.pub`) and paste it into the "Key" field.
   - Click "Add SSH key."

4. **Use SSH for Git Operations:**
   - In your terminal, change the remote URL of your repository to use the SSH URL instead of the HTTPS URL:
     ```sh
     git remote set-url origin git@github.com:username/repo.git
     ```
   - Now, you can use the SSH key for authentication when pushing or pulling from the repository.

Using either a personal access token or an SSH key is more secure and also helps you avoid entering your password repeatedly. Additionally, these methods are compatible with various Git clients and IDEs, so you'll have a seamless experience across different tools.

**Comment by queenofcorgis** [comment]
Thanks @arafajs @w8ste for jumping in to help. Don't forget @MvuleniNofemele to mark the answer if either of these helped solve your question ✅
